Do not create origin directory or origin file for LocalStorage prewarming

szewai opened the pull request:
#### f0e6ad03798874ee0b2520dca4ec7c47f9c2e556
<pre>
Do not create origin directory or origin file for LocalStorage prewarming
<a href="https://bugs.webkit.org/show_bug.cgi?id=242064">https://bugs.webkit.org/show_bug.cgi?id=242064</a>

Reviewed by Chris Dumez.

We currently prewarm LocalStorage (try opening the LocalStorage database and fetch items before the API is called) for
performance, and we create origin directory and origin file when OriginStorageManager is created. This means we would
create the directory for each origin visited, but we should create the directory when it is needed, i.e. when there will
be content inside it; otherwise the origin file and directory are useless and will take up user&apos;s disk space.

API test: WebKit.CreateOriginFileWhenNeeded

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::writeOriginToFile):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::localOriginStorageManager):
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::setItem):
(WebKit::NetworkStorageManager::removeItem):
(WebKit::NetworkStorageManager::clear):
(WebKit::writeOriginToFileIfNecessary): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
(WebKit::OriginStorageManager::didWriteOriginToFile const):
(WebKit::OriginStorageManager::markDidWriteOriginToFile):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/251997@main">https://commits.webkit.org/251997@main</a>
</pre>

cdumez commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
```diff
@@ -65,6 +65,8 @@ class OriginStorageManager {
     void deleteData(OptionSet<WebsiteDataType>, WallTime);
     void moveData(OptionSet<WebsiteDataType>, const String& localStoragePath, const String& idbStoragePath);
+    bool didWriteOriginToFile() const { return m_didWriteOriginToFile; }
+    void setDidWriteOriginToFile(bool didWriteOriginToFile) { m_didWriteOriginToFile = didWriteOriginToFile; }
```
This only ever sets it to true so I think this function shouldn't have a parameter:
`void setDidWriteOriginToFile() { m_didWriteOriginToFile = didWriteOriginToFile; }`
or
`void markDidWriteOriginToFile() { m_didWriteOriginToFile = didWriteOriginToFile; }`

szewai replied:
Sure.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+        hasQuotaError = (result.error() == StorageError::QuotaExceeded);
     completionHandler(hasQuotaError);
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```
Looks a little risky to call this *after* calling the completion handler? Are we sure the completion handler cannot cause any work that would require the origin to be written?

szewai replied:
This is a message handler, so the completion handler is just sending reply

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+    storageArea->removeItem(connection.uniqueID(), implIdentifier, WTFMove(key), WTFMove(urlString));
     completionHandler();
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```
ditto.

szewai replied:
ditto.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+    storageArea->clear(connection.uniqueID(), implIdentifier, WTFMove(urlString));
     completionHandler();
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```
ditto.

szewai replied:
ditto.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -244,14 +244,29 @@ static String originFilePath(const String& directory)
+    if (originDirectory.isEmpty())
+        return;
+
+    if (!condition)
```
This looks wrong? It checks if condition is null but doesn't actually call the function?
If I'm right, I guess it means testing is insufficient?

As a side note, since condition is run synchronously, we can probably pass it as a `const Function<bool>()>&` instead of a rvalue reference.

szewai replied:
oooops. I should add a test case to verify there is origin file if the condition is true. Will change to const Function<bool>()>& too.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -261,6 +276,12 @@ OriginStorageManager& NetworkStorageManager::localOriginStorageManager(const Web
         return makeUnique<OriginStorageManager>(quota, WTFMove(increaseQuotaFunction), WTFMove(originDirectory), WTFMove(localStoragePath), WTFMove(idbStoragePath), WTFMove(cacheStoragePath), m_shouldUseCustomPaths);
     }).iterator->value;
+
+    writeOriginToFileIf(origin, [&]() {
```
nit: () is not needed.

szewai replied:
Will remove

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -795,8 +807,13 @@ void NetworkStorageManager::connectToStorageArea(IPC::Connection& connection, We
-        return completionHandler(resultIdentifier, storageArea->allItems(), StorageAreaBase::nextMessageIdentifier());
+    if (auto storageArea = m_storageAreaRegistry->getStorageArea(resultIdentifier)) {
+        completionHandler(resultIdentifier, storageArea->allItems(), StorageAreaBase::nextMessageIdentifier());
+        writeOriginToFileIf(origin, [&]() {
```
ditto.

szewai replied:
Will remove

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+    if (!storageArea)
+        return completionHandler(hasQuotaError);
 
+    auto result = storageArea->setItem(connection.uniqueID(), implIdentifier, String { key }, WTFMove(value), WTFMove(urlString));
```
nit: Why isn't this using `WTFMove(key)` ?

szewai replied:
I don't remember; but should be fine to WTFMove...

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+        hasQuotaError = (result.error() == StorageError::QuotaExceeded);
     completionHandler(hasQuotaError);
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```
nit: () is not necessary.

szewai replied:
Will remove

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
+    storageArea->removeItem(connection.uniqueID(), implIdentifier, WTFMove(key), WTFMove(urlString));
     completionHandler();
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
```
ditto.

szewai replied:
Will remove

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -244,14 +244,29 @@ static String originFilePath(const String& directory)
     return FileSystem::pathByAppendingComponent(directory, OriginStorageManager::originFileIdentifier());
 }
 
+void NetworkStorageManager::writeOriginToFileIf(const WebCore::ClientOrigin& origin, const Function<bool()>& condition)
```
Why aren't we doing the condition check at call sites?

szewai replied:
Because the first checks in the functions are cheaper; if we can return early here, we don't need to call condition() which may involve file operation.

cdumez replied:
Oh I see, SQLiteStorageArea::isEmpty() is not trivial indeed. It is not obvious from the name.

My comment about code duplication still stands though.

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -244,14 +244,29 @@ static String originFilePath(const String& directory)
+    if (originDirectory.isEmpty())
+        return;
+
+    if (!condition())
```
Why do we have to do the condition check so late? Why do we bother running all this code above if the condition will return false anyway?

szewai replied:
Because the previous checks are cheaper... SQLiteStorageArea::isEmpty() may have file operations

cdumez commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -857,33 +874,50 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
     completionHandler();
+
+    writeOriginToFileIf(storageArea->origin(), [&]() {
+        return storageArea->type() == StorageAreaBase::Type::SQLite && !storageArea->isEmpty();
```
This is repeated so many times, this should be factored in its own function to avoid repetition, e.g.:
```
if (shouldWriteOriginToFile(storageArea))
    writeOriginToFile(storageArea->origin());
```

szewai replied:
Sure

cdumez commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -68,6 +68,7 @@ SQLiteStorageArea::SQLiteStorageArea(unsigned quota, const WebCore::ClientOrigin
     , m_queue(WTFMove(workQueue))
     , m_cachedStatements(static_cast<size_t>(StatementType::Invalid))
 {
+    WTFLogAlways("sihuil: [%p]SQLiteStorageArea::SQLiteStorageArea m_path[%s]", this, m_path.utf8().data());
```
This shouldn't be here.

cdumez reviewed (approved):
LGTM

szewai committed:
Do not create origin directory or origin file for LocalStorage prewarming
https://bugs.webkit.org/show_bug.cgi?id=242064

Reviewed by Chris Dumez.

We currently prewarm LocalStorage (try opening the LocalStorage database and fetch items before the API is called) for
performance, and we create origin directory and origin file when OriginStorageManager is created. This means we would
create the directory for each origin visited, but we should create the directory when it is needed, i.e. when there will
be content inside it; otherwise the origin file and directory are useless and will take up user's disk space.

API test: WebKit.CreateOriginFileWhenNeeded

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::writeOriginToFile):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::localOriginStorageManager):
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::setItem):
(WebKit::NetworkStorageManager::removeItem):
(WebKit::NetworkStorageManager::clear):
(WebKit::writeOriginToFileIfNecessary): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
(WebKit::OriginStorageManager::didWriteOriginToFile const):
(WebKit::OriginStorageManager::markDidWriteOriginToFile):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(TEST):

Canonical link: https://commits.webkit.org/251997@main
